src/Native/Runtime/amd64/StubDispatch.S

```diff
@@ -0,0 +1,171 @@
+
+LEAF_ENTRY RhpInterfaceDispatch\entries, _TEXT
+
+        // r10 currently contains the indirection cell address.
+        // load r11 to point to the cache block.
+        mov     r11, [r10 + OFFSETOF__InterfaceDispatchCell__m_pCache]
+
+        // Load the EEType from the object instance in rdi.
+        mov     rax, [rdi]
+
+        // Point r11 at the address of the first cache entry
+        // Switch to ATT syntax because add r11, OFFSETOF__InterfaceDispatchCache__m_rgEntries
+        // refuses to treat symbols as immediate values.
+        .att_syntax
+        addq     $OFFSETOF__InterfaceDispatchCache__m_rgEntries, %r11
+        .intel_syntax noprefix
+
+        // For each entry in the cache, see if its EEType type matches the EEType in rax.
+        // If so, call the second cache entry.  If not, skip the InterfaceDispatchCacheEntry.
+        .rept \entries
+            cmp     rax, [r11]
+            je      DispatchCall\entries
+
+            .att_syntax
+            addq    $SIZEOF__InterfaceDispatchCacheEntry, %r11
+            .intel_syntax noprefix
+        .endr
+
+        // r10 still contains the the indirection cell address.
+
+        jmp     C_FUNC(RhpInterfaceDispatchSlow)
+DispatchCall\entries:
+        add     r11, 0x8h
+        mov     rax, [r11]
+        jmp     rax
+LEAF_END RhpInterfaceDispatch\entries, _TEXT
+
+.endm // DEFINE_INTERFACE_DISPATCH_STUB
+
+
+
+// Define all the stub routines we currently need.
+//
+// The mrt100dbi requires these be exported to identify mrt100 code that dispatches back into managed.
+// If you change or add any new dispatch stubs, please also change slr.def and dbi\process.cpp CordbProcess::GetExportStepInfo
+//
+DEFINE_INTERFACE_DISPATCH_STUB 1
+DEFINE_INTERFACE_DISPATCH_STUB 2
+DEFINE_INTERFACE_DISPATCH_STUB 4
+DEFINE_INTERFACE_DISPATCH_STUB 8
+DEFINE_INTERFACE_DISPATCH_STUB 16
+DEFINE_INTERFACE_DISPATCH_STUB 32
+DEFINE_INTERFACE_DISPATCH_STUB 64
+
+// Initial dispatch on an interface when we don't have a cache yet.
+LEAF_ENTRY RhpInitialInterfaceDispatch, _TEXT
+ALTERNATE_ENTRY RhpInitialDynamicInterfaceDispatch
+
+        // Just tail call to the cache miss helper.
+        jmp     C_FUNC(RhpInterfaceDispatchSlow)
+
+LEAF_END RhpInitialInterfaceDispatch, _TEXT
+
+// Cache miss case, call the runtime to resolve the target and update the cache.
+NESTED_ENTRY RhpInterfaceDispatchSlow, _TEXT, NoHandler
+
+        .equ RIDS_ReservedStack, 30h + 50h + 80h + 8h     // Scratch space, transition frame, xmm registers and padding
+
+        alloc_stack         RIDS_ReservedStack
+
+        // Preserve the argument registers in the scratch space across the helper call. Note that we depend on these
+        // registers (which may contain GC references) being spilled before we build the PInvokeTransitionFrame below
+        // due to the way we build a stack range to report to the GC conservatively during a collection.
+        mov     [rsp + 8*0], rdi
```

Comment:
It seems that using push to store the registers would have better perf. You can use the PUSH_ARGUMENT_REGISTERS / POP_ARGUMENT_REGISTERS. Using the macro also has the benefit of the .cfi annotation that's in it and lets the unwinder understand where the registers are saved.